[16/N] Add _allgather_base custom op with CPU/CUDA implementation

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88330 Allow Process Group to support multiple backends and move PG specfic implementations to backend class
* #87088 Remove ProcessGroupRoundRobin
* #88903 [17/N] Add _reduce_scatter_base custom op with CPU/CUDA implementation
* **#88889 [16/N] Add _allgather_base custom op with CPU/CUDA implementation**
* #88846 [15/N] Add allreduce_coalesced custom op with CPU/CUDA implementations
* #88351 [14/N] Refactor _new_process_group_helper() to remove repeated code

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D41227739](https://our.internmc.facebook.com/intern/diff/D41227739)

H-Huang committed:
[16/N] Add _allgather_base custom op with CPU/CUDA implementation

[ghstack-poisoned]

H-Huang commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1187,7 +1187,13 @@ that adds a prefix to each key inserted to the store.
-              &::c10d::ProcessGroup::_allgather_base,
+              [](const c10::intrusive_ptr<::c10d::ProcessGroup>& self,
+                 at::Tensor& output_tensor,
+                 at::Tensor& input_tensor,
```
i think this can be const &

kwen2501 replied:
I had a look at `init.cpp`, it seems the convention has not been using const. Not sure why. Can go with convention I guess.

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2958,6 +2958,23 @@ def test_collectives(self):
+        device = "cuda"
+        tensor = torch.ones(10, 10, device=torch.device(device))
+        output_tensor = torch.zeros(10, 10, device=torch.device(device))
+        dist._all_gather_base(output_tensor, tensor)
```
Please use `all_gather_into_tensor` at Python level.
We will also rename `_all_gather_base` at the binding and C++ levels to `all_gather_into_tensor` when we introduce extension-breaking changes in 2.0.

H-Huang replied:
good point! thanks

kwen2501 reviewed (approved):
LGTM.

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41227739).

H-Huang committed:
Update on "[16/N] Add _allgather_base custom op with CPU/CUDA implementation"


Differential Revision: [D41227739](https://our.internmc.facebook.com/intern/diff/D41227739)

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge
